gui/packages/desktop/src/main/index.js

```diff
@@ -190,13 +210,111 @@ const ApplicationMain = {
     window.loadFile(path.resolve(path.join(__dirname, '../renderer/index.html')));
   },
 
+  async _onDaemonConnected() {
+    this._connectedToDaemon = true;
+
+    // reset the reconnect backoff when connection established.
+    this._reconnectBackoff.reset();
+
+    if (this._windowController) {
+      this._windowController.send('daemon-connected');
+    }
+
+    const stateListener = new SubscriptionListener(
+      (newState: TunnelStateTransition) => {
+        this._onStateChange(newState);
+      },
+      (error: Error) => {
+        log.error(`Cannot deserialize the new state: ${error.message}`);
+      },
+    );
+
+    const settingsListener = new SubscriptionListener(
+      (newSettings: Settings) => {
+        this._onSettingsChange(newSettings);
+      },
+      (error: Error) => {
+        log.error(`Cannot deserialize the new settings: ${error.message}`);
+      },
+    );
+
+    try {
+      await Promise.all([
+        this._daemonRpc.subscribeStateListener(stateListener),
+        this._daemonRpc.subscribeSettingsListener(settingsListener),
+      ]);
+    } catch (error) {
+      log.error(`Failed to subscribe: ${error.message}`);
+    }
+  },
+
+  _onDaemonDisconnected(error: ?Error) {
+    // recover connection on error
+    if (error) {
+      log.debug(`Lost connection to daemon: ${error.message}`);
+
+      this._reconnectBackoff.attempt(() => {
+        this._connectToDaemon();
+      });
+    } else {
+      log.info(`Disconnected from the daemon`);
+    }
+
+    this._connectedToDaemon = false;
+
+    if (this._windowController) {
+      this._windowController.send('daemon-disconnected', error ? error.message : null);
+    }
+  },
+
+  _connectToDaemon() {
+    this._daemonRpc.connect({ path: DAEMON_RPC_PATH });
+  },
+
+  _onStateChange(newState: TunnelStateTransition) {
+    if (this._windowController) {
+      this._windowController.send('state-changed', newState);
+    }
+  },
+
+  _onSettingsChange(newSettings: Settings) {
+    if (this._windowController) {
+      this._windowController.send('settings-changed', newSettings);
+    }
+  },
+
   _registerWindowListener(windowController: WindowController) {
     const window = windowController.window;
 
     window.on('show', () => window.webContents.send('window-shown'));
   },
 
   _registerIpcListeners() {
+    ipcMain.on('daemon-rpc-call', async (event, id: string, method: string, payload: any) => {
+      log.debug(`Got daemon-rpc-call: ${id} ${method} ${payload}`);
+
+      try {
+        // $FlowFixMe: flow does not like index accessors.
+        const result = await this._daemonRpc[method](payload);
+
+        log.debug(`Reply to ${id} ${method} with success: ${JSON.stringify(result)}`);
+        event.sender.send(`daemon-rpc-reply-${id}`, result);
+      } catch (error) {
+        log.debug(`Reply to ${id} ${method} with error: ${error.message}`);
+        event.sender.send(`daemon-rpc-reply-${id}`, undefined, {
```

Comment:
Electron IPC uses JSON transport so there is no way to serialise `Error` subclasses. Instead I save the class name and attach the payload with `message` and any custom fields collected with help of `JSON.stringify`. Then the error is being re-constructed on the renderer side.